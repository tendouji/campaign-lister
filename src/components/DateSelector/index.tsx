import React from 'react';
import { connect } from 'react-redux';
import styled from "styled-components";
import DatePicker from "react-datepicker";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCalendarCheck } from '@fortawesome/free-solid-svg-icons/faCalendarCheck'
import { faArrowsAltH } from '@fortawesome/free-solid-svg-icons/faArrowsAltH'
import { faTimes } from "@fortawesome/free-solid-svg-icons/faTimes";
import { colors, layout, styles } from '../../constants/ui';
import { setSnackBarDisplay, setDateRange } from "../../actions";
import { DateRangeData, ReducerStateData, SnackBarData } from '../../models';

import "react-datepicker/dist/react-datepicker.css";


type DateSelectorProps = {
    [key:string]: any
}

export class DateSelector extends React.Component<DateSelectorProps, any> {
    constructor(props: DateSelectorProps) {
        super(props);

        this.setDate = this.setDate.bind(this);
        this.onClearDateRange = this.onClearDateRange.bind(this);
        this.onDateFilterClick = this.onDateFilterClick.bind(this);

        const dateNow: Date = new Date();

        this.state = {
            dateNow,
            startDate: dateNow,
            endDate: dateNow,
        };
    }

    componentDidUpdate(prevProps: any) {
        const { appState } = this.props;
        const { dateRange } = appState;

        if(prevProps.appState.dateRange.startDate !== dateRange.startDate) {
            if(!dateRange.startDate) {
                // NOTE: startDate is null
                this.onClearDateRange();
            }
        }
    }

    onDateFilterClick(e?: React.MouseEvent | undefined) {
        const { startDate, endDate } = this.state;

        if(startDate > endDate) {
            const { setSnackBarDisplay } = this.props;
            setSnackBarDisplay({
                show: true,
                message: 'Start date cannot be later than end date.'
            });
            return;
        }

        const { setDateRange } = this.props;
        setDateRange({
            startDate,
            endDate,
        });
    };

    setDate(dateVal: Date, type: string) {
        this.setState({ [type]: dateVal }, () => {
            const { startDate, endDate } = this.state;
            if(startDate > endDate) {
                this.setState({ endDate: startDate }, () => {
                    this.onDateFilterClick();
                });
            } else {
                this.onDateFilterClick();
            }
        });
    };

    onClearDateRange() {
        const { dateNow } = this.state;
        const { setDateRange } = this.props;
        setDateRange({ startDate: null, endDate: null });
        this.setState({
            startDate: dateNow,
            endDate: dateNow
        });
    };

    render() {
        const { startDate, endDate } = this.state;
        const { appState } = this.props;
        const { dateRange } = appState;

        return (
            <DateSelectorWrapper className="date-selector">
                <div className="date-area">
                    <div className="date-col start-date">
                        <div className="label">Start Date</div>
                        <DatePicker
                            id="startDatePicker"
                            selected={startDate}
                            onChange={(date: Date) => this.setDate(date, 'startDate')}
                            placeholderText='Select your start date'
                            popperPlacement="bottom"
                            todayButton="Today"
                            className="datepicker-input"
                        />
                    </div>
                    <div className="icon-range">
                        <FontAwesomeIcon icon={faArrowsAltH} />
                    </div>
                    <div className="date-col end-date">
                        <div className="label">End Date</div>
                        <DatePicker
                            id="endDatePicker"
                            selected={endDate < startDate ? startDate : endDate}
                            onChange={(date: Date) => this.setDate(date, 'endDate')}
                            minDate={startDate}
                            placeholderText='Select your end date'
                            popperPlacement="bottom"
                            todayButton="Today"
                            className="datepicker-input"
                        />
                    </div>
                </div>
                { !!dateRange.startDate || !!dateRange.endDate
                    ? <button id="clearDateRange" onClick={this.onClearDateRange}><FontAwesomeIcon icon={faTimes} /></button>
                    : <button id="goFilter" onClick={this.onDateFilterClick}><FontAwesomeIcon icon={faCalendarCheck} /></button>
                }
            </DateSelectorWrapper>
        );
    }
}

const mapStateToProps = (state: ReducerStateData) => {
    return ({ appState: state.appState })
};

const mapDispatchToProps = (dispatch: any) => ({
    setDateRange: (data: DateRangeData) => dispatch(setDateRange(data)),
    setSnackBarDisplay: (data: SnackBarData) => dispatch(setSnackBarDisplay(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(DateSelector);



const DateSelectorWrapper = styled.div`
    display: flex;
    width: 100%;
    align-items: center;
    
    & .date-area {
        display: flex;
        width: calc(100% - ${layout.filterBarHeight});
        padding: 0 ${layout.largeGap};
        box-sizing: border-box;
        align-items: center;
        justify-content: space-between;
    }
    
    & .date-col {
        box-sizing: border-box;
        font-size: ${layout.fontSizeStandard};
        
        & > .label {
            margin-bottom: ${layout.xSmallGap};
            font-size: ${layout.fontSizeSmall};
            font-weight: 500;
        }
    }
    
    & .icon-range {
        padding: 1.4rem 3px 0;
    }
    
    & > button {
        ${styles.filterBarButtonStyle};
    }
    
    & .react-datepicker__triangle {
        left: 50%;
    }
    
    & .react-datepicker__input-container {
        & > input[type="text"] {
            width: 6rem;
            padding: ${layout.xSmallGap} ${layout.smallGap};
            border: ${styles.lineStyle};
            font-size: ${layout.fontSizeStandard};
            text-align: center;
            background-color: ${colors.lightGray};
            box-sizing: border-box;
            
            &:focus {
                outline: none;
            }
        }
    }
`;