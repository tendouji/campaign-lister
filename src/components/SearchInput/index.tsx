import React from 'react';
import { connect } from 'react-redux';
import styled from "styled-components";
import { colors, layout, styles} from '../../constants/ui';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSearch } from '@fortawesome/free-solid-svg-icons/faSearch'
import { faTimes } from '@fortawesome/free-solid-svg-icons/faTimes'
import { setSearchKey, setSnackBarDisplay } from "../../actions";
import { ReducerStateData, SnackBarData } from '../../models';
import { config } from "../../constants/config";


type SearchInputState = {
    searchKeyValue: string
}

class SearchInputBase extends React.Component<any, SearchInputState> {
    private searchKeyRef: React.RefObject<HTMLInputElement>;
    private searchKeyTimer: ReturnType<typeof setTimeout> = 0;

    constructor(props: any) {
        super(props);

        this.searchKeyRef = React.createRef();
        this.onSearchKeyChange = this.onSearchKeyChange.bind(this);

        this.state = {
            searchKeyValue: ''
        }
    }

    componentDidUpdate(prevProps: any) {
        const { appState } = this.props;

        if(!!prevProps.appState && !!appState) {
            if(prevProps.appState.searchKey !== appState.searchKey) {
                this.setState({ searchKeyValue: appState.searchKey });
            }
        }
    }

    onSearchClick = (e?: React.MouseEvent | undefined) => {
        const searchKeyElm = this.searchKeyRef.current;
        if(!!searchKeyElm) {
            const searchVal = searchKeyElm.value;

            if(searchVal.length < config.minSearchLength) {
                if(!!e) {
                    // Note: if e doesn't exist, it is called from onChange event
                    const { setSnackBarDisplay } = this.props;
                    setSnackBarDisplay({
                        show: true,
                        message: `Search length is too short. Please enter at least ${config.minSearchLength} characters`
                    });
                }
                return;
            }

            const { setSearchKey } = this.props;
            setSearchKey(searchVal);
        }
    };

    onSearchKeyChange = () => {
        const searchKeyElm = this.searchKeyRef.current;
        if(!!searchKeyElm) {
            this.setState({searchKeyValue: searchKeyElm.value});

            if(searchKeyElm.value === '') {
                this.onClearSearchKey();
            }
        }

        clearTimeout(this.searchKeyTimer);
        this.searchKeyTimer = setTimeout(() => {
            this.onSearchClick();
        }, 500)
    };

    onClearSearchKey = () => {
        const { setSearchKey } = this.props;
        setSearchKey('');
    };

    render() {
        const { searchKeyValue } = this.state;
        const { appState } = this.props;

        return (
            <SearchWrapper className="search-input">
                <input type="text"
                    id="searchKey"
                    ref={this.searchKeyRef}
                    aria-label="Search Input"
                    autoComplete="off"
                    spellCheck={false}
                    placeholder="Search Campaign"
                    onChange={this.onSearchKeyChange}
                    value={searchKeyValue}
                />
                { !!appState && appState.searchKey !== ''
                    ? <button onClick={this.onClearSearchKey}><FontAwesomeIcon icon={faTimes} /></button>
                    : <button onClick={this.onSearchClick}><FontAwesomeIcon icon={faSearch} /></button>
                }
            </SearchWrapper>
        )
    }
}

const mapStateToProps = (state: ReducerStateData) => {
    return ({ appState: state.appState })
};

const mapDispatchToProps = (dispatch: any) => ({
    setSearchKey: (data: string) => dispatch(setSearchKey(data)),
    setSnackBarDisplay: (data: SnackBarData) => dispatch(setSnackBarDisplay(data)),
});

const SearchInput = connect(mapStateToProps, mapDispatchToProps)(SearchInputBase);

export default SearchInput;


const SearchWrapper = styled.div`
    display: flex;
    width: 100%;
    align-item: center;
    
    & #searchKey {
        width: calc(100% - ${layout.filterBarHeight});
        padding: ${layout.largeGap};
        border: 0;
        font-size: ${layout.fontSizeStandard};
        
        &:focus {
            outline: none;
        }
    }
    
    & > button {
        ${styles.filterBarButtonStyle};
    }
`;