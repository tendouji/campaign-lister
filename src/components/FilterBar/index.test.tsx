import React from 'react';
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";
import { FilterBar } from './';
import SearchInput from "../SearchInput";
import DateSelector from "../DateSelector";


describe('<FilterBar />', () => {
    const wrapper = shallow(<FilterBar />);

    it('renders without crashing', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders the correct child components', () => {
        expect(wrapper.find(SearchInput)).toHaveLength(1);
        expect(wrapper.find(DateSelector)).toHaveLength(1);
    });
});