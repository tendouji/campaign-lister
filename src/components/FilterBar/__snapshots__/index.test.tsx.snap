// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterBar /> renders without crashing 1`] = `
<styled.div
  className="app-filter"
>
  <div
    className="column search-panel"
  >
    <Connect(SearchInput) />
  </div>
  <div
    className="column date-panel"
  >
    <Connect(DateSelector) />
  </div>
</styled.div>
`;
